### ISSUE TITLE
SSL: CERTIFICATE_VERIFY_FAILED error with `boto3`

### ISSUE THREAD

**Issue by prageeth** [opravdin/weback-unofficial#9]
I tried to use the `weback HACS` integration and keep receving this error:

```
weback: Error on device update!
Traceback (most recent call last):
  File "/srv/homeassistant/lib/python3.8/site-packages/urllib3/connectionpool.py", line 699, in urlopen
    httplib_response = self._make_request(
  File "/srv/homeassistant/lib/python3.8/site-packages/urllib3/connectionpool.py", line 382, in _make_request
    self._validate_conn(conn)
  File "/srv/homeassistant/lib/python3.8/site-packages/urllib3/connectionpool.py", line 1010, in _validate_conn
    conn.connect()
  File "/srv/homeassistant/lib/python3.8/site-packages/urllib3/connection.py", line 411, in connect
    self.sock = ssl_wrap_socket(
  File "/srv/homeassistant/lib/python3.8/site-packages/urllib3/util/ssl_.py", line 428, in ssl_wrap_socket
    ssl_sock = _ssl_wrap_socket_impl(
  File "/srv/homeassistant/lib/python3.8/site-packages/urllib3/util/ssl_.py", line 472, in _ssl_wrap_socket_impl
    return ssl_context.wrap_socket(sock, server_hostname=server_hostname)
  File "/usr/local/lib/python3.8/ssl.py", line 500, in wrap_socket
    return self.sslsocket_class._create(
  File "/usr/local/lib/python3.8/ssl.py", line 1040, in _create
    self.do_handshake()
  File "/usr/local/lib/python3.8/ssl.py", line 1309, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1123)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/srv/homeassistant/lib/python3.8/site-packages/botocore/httpsession.py", line 254, in send
    urllib_response = conn.urlopen(
  File "/srv/homeassistant/lib/python3.8/site-packages/urllib3/connectionpool.py", line 755, in urlopen
    retries = retries.increment(
  File "/srv/homeassistant/lib/python3.8/site-packages/urllib3/util/retry.py", line 506, in increment
    raise six.reraise(type(error), error, _stacktrace)
  File "/srv/homeassistant/lib/python3.8/site-packages/urllib3/packages/six.py", line 734, in reraise
    raise value.with_traceback(tb)
  File "/srv/homeassistant/lib/python3.8/site-packages/urllib3/connectionpool.py", line 699, in urlopen
    httplib_response = self._make_request(
  File "/srv/homeassistant/lib/python3.8/site-packages/urllib3/connectionpool.py", line 382, in _make_request
    self._validate_conn(conn)
  File "/srv/homeassistant/lib/python3.8/site-packages/urllib3/connectionpool.py", line 1010, in _validate_conn
    conn.connect()
  File "/srv/homeassistant/lib/python3.8/site-packages/urllib3/connection.py", line 411, in connect
    self.sock = ssl_wrap_socket(
  File "/srv/homeassistant/lib/python3.8/site-packages/urllib3/util/ssl_.py", line 428, in ssl_wrap_socket
    ssl_sock = _ssl_wrap_socket_impl(
  File "/srv/homeassistant/lib/python3.8/site-packages/urllib3/util/ssl_.py", line 472, in _ssl_wrap_socket_impl
    return ssl_context.wrap_socket(sock, server_hostname=server_hostname)
  File "/usr/local/lib/python3.8/ssl.py", line 500, in wrap_socket
    return self.sslsocket_class._create(
  File "/usr/local/lib/python3.8/ssl.py", line 1040, in _create
    self.do_handshake()
  File "/usr/local/lib/python3.8/ssl.py", line 1309, in do_handshake
    self._sslobj.do_handshake()
urllib3.exceptions.SSLError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1123)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/srv/homeassistant/lib/python3.8/site-packages/homeassistant/helpers/entity_platform.py", line 359, in _async_add_entity
    await entity.async_device_update(warning=False)
  File "/srv/homeassistant/lib/python3.8/site-packages/homeassistant/helpers/entity.py", line 466, in async_device_update
    await task
  File "/usr/local/lib/python3.8/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/home/homeassistant/.homeassistant/custom_components/weback/vacuum.py", line 64, in update
    self.device.update()
  File "/srv/homeassistant/lib/python3.8/site-packages/weback_unofficial/client.py", line 165, in update
    shadow = self.client.get_device_shadow(self.name)
  File "/srv/homeassistant/lib/python3.8/site-packages/weback_unofficial/client.py", line 102, in get_device_shadow
    resp = client.get_thing_shadow(thingName=device_name)
  File "/srv/homeassistant/lib/python3.8/site-packages/botocore/client.py", line 357, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/srv/homeassistant/lib/python3.8/site-packages/botocore/client.py", line 662, in _make_api_call
    http, parsed_response = self._make_request(
  File "/srv/homeassistant/lib/python3.8/site-packages/botocore/client.py", line 682, in _make_request
    return self._endpoint.make_request(operation_model, request_dict)
  File "/srv/homeassistant/lib/python3.8/site-packages/botocore/endpoint.py", line 102, in make_request
    return self._send_request(request_dict, operation_model)
  File "/srv/homeassistant/lib/python3.8/site-packages/botocore/endpoint.py", line 136, in _send_request
    while self._needs_retry(attempts, operation_model, request_dict,
  File "/srv/homeassistant/lib/python3.8/site-packages/botocore/endpoint.py", line 253, in _needs_retry
    responses = self._event_emitter.emit(
  File "/srv/homeassistant/lib/python3.8/site-packages/botocore/hooks.py", line 356, in emit
    return self._emitter.emit(aliased_event_name, **kwargs)
  File "/srv/homeassistant/lib/python3.8/site-packages/botocore/hooks.py", line 228, in emit
    return self._emit(event_name, kwargs)
  File "/srv/homeassistant/lib/python3.8/site-packages/botocore/hooks.py", line 211, in _emit
    response = handler(**kwargs)
  File "/srv/homeassistant/lib/python3.8/site-packages/botocore/retryhandler.py", line 183, in __call__
    if self._checker(attempts, response, caught_exception):
  File "/srv/homeassistant/lib/python3.8/site-packages/botocore/retryhandler.py", line 250, in __call__
    should_retry = self._should_retry(attempt_number, response,
  File "/srv/homeassistant/lib/python3.8/site-packages/botocore/retryhandler.py", line 277, in _should_retry
    return self._checker(attempt_number, response, caught_exception)
  File "/srv/homeassistant/lib/python3.8/site-packages/botocore/retryhandler.py", line 316, in __call__
    checker_response = checker(attempt_number, response,
  File "/srv/homeassistant/lib/python3.8/site-packages/botocore/retryhandler.py", line 222, in __call__
    return self._check_caught_exception(
  File "/srv/homeassistant/lib/python3.8/site-packages/botocore/retryhandler.py", line 359, in _check_caught_exception
    raise caught_exception
  File "/srv/homeassistant/lib/python3.8/site-packages/botocore/endpoint.py", line 200, in _do_get_response
    http_response = self._send(request)
  File "/srv/homeassistant/lib/python3.8/site-packages/botocore/endpoint.py", line 269, in _send
    return self.http_session.send(request)
  File "/srv/homeassistant/lib/python3.8/site-packages/botocore/httpsession.py", line 281, in send
    raise SSLError(endpoint_url=request.url, error=e)
botocore.exceptions.SSLError: SSL validation failed for https://data.iot.eu-central-1.amazonaws.com/things/ty-x500-cc-50-e3-23-06-11/shadow [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1123)
```

This seems to be related to the Certs used by `data.iot.xxx.amazonaws.com` being deprecated; issue reported in the `boto3` project: https://github.com/boto/boto3/issues/2686

The workaround seems to be that we have to use the `iot:Data-ATS` endpoint, and in order to do that, we need to explicitly specify that when initialising the `client` here: https://github.com/opravdin/weback-unofficial/blob/5afba4546ba3e1a387bf7f3ce6e1e8307fbcb6b6/weback_unofficial/client.py#L101

**Comment by pops106** [comment]
Just upgraded to new version of hassio and found the same issue, robot vacuum has been working so I guess this would of showed up with just a reboot.

If I open a webpage to the URL in the config it looks like a dead Symantec cert error.

NET::ERR_CERT_SYMANTEC_LEGACY

**Comment by prageeth** [comment]
@opravdin Hi, any chance this will be fixed soon?

**Comment by pops106** [comment]
This might be helpful, seems to explain the issue pretty well but it is a bit beyond me.

https://docs.aws.amazon.com/iot/latest/developerguide/iot-connect-devices.html

**Comment by pops106** [comment]
Interestingly, I just set this up in Ubuntu 20 using the instructions provided and it connects perfectly and I can control the robot. So maybe this is something which has updated in hassio.

I did read a thing about upgrading python removed support for some older certificates.

Just updating everything in Ubuntu now to see if I can break it.

**Comment by opravdin** [comment]
Hello everyone. Thank you for opening issue, I will fix this in a couple of days

**Comment by mattes66** [comment]
I had the same issue. API just work and after installation paho-mqtt it stopped working with the same error message. 
My workaround was simple: I just downgraded certifi from version 2020.12.5 to 2020.11.8
conda install certifi=2020.11.8
But this is no long term solution since the problem will surely reappear after next update.

**Comment by a8ree** [comment]
> I had the same issue. API just work and after installation paho-mqtt it stopped working with the same error message.
> My workaround was simple: I just downgraded certifi from version 2020.12.5 to 2020.11.8
> conda install certifi=2020.11.8
> But this is no long term solution since the problem will surely reappear after next update.

Sorry, I don't understand how to implement this workaround. Can you point me in the right direction @mattes66

**Comment by prageeth** [comment]
Hi @opravdin, hope you had a good new year. I was wondering if you had an update on this?

**Comment by vityz** [comment]
@opravdin Привет. Я так понимаю сейчас не работает эта интеграция из-за сертификатов ? Устанавливать смысла нет ?

**Comment by opravdin** [comment]
Update here: I will try to reverse engineer newer version of weback app. Amazon suggests users to migrate on client-specific (in this case client is a weback app developer team) endpoints. Since I am not a weback app developer, I can't just jump in AWS account and grab weback's endpoint name.

**Comment by pops106** [comment]
I thought I would try and help but it is way beyond me...

I found this but I am sure you need more than this;

File was AWSIotClient.java
setEndpoint("iot.us-east-1.amazonaws.com");

**Comment by bishely** [comment]
> I will try to reverse engineer newer version of weback app.

Is there a new version? I haven't seen an update on my phone (android). I'm running into this same error trying to deploy a script to a Raspberry Pi. Strangely, it worked (and still works) on the Mac where I wrote the script, so I wondered if it might just be the certificate not existing on some devices/OSes?

I've tried manually adding the aws endpoint's cert (in my case, data.iot.eu-central-1.amazonaws.com) to the Pi, both at the OS level (save to /usr/local/share/ca-certificates/ and run update-ca-certificates) and by appending it to Certifi's cacert.pem, but I still get the verify failed error.

**Comment by timdreier** [comment]
I don’t have my MacBook with me in the moment and haven’t looked in the code, but might this help?:
https://www.shellhacks.com/aws-cli-ssl-validation-failed-solved/

**Comment by timdreier** [comment]
Hey @opravdin, I think I fixed the issue (see my [pull request](https://github.com/opravdin/weback-unofficial/pull/12)). I've never done something in python before, so I hope everything is fine.

The issue seems to be the new version of certifi which drops the certificates amazon used on this server. I used the function described [here](https://github.com/boto/boto3/issues/2686) to obtain the client-specific endpoint url.

I've not done much testing since I'm not home in the moment, but I was able to obtain the state of my Tesvor after this fix. I really hope this helps / fixes this issue!
